File: frontend/src/types/eval-explorer.ts
// ---------------------------------------------------------------------------
// Assessment & Trace types
// ---------------------------------------------------------------------------

export interface AssessmentDetail {
  name: string;
  raw_value: string | number | boolean;
  normalized_score: number | null;
  passed: boolean | null;
  rationale: string | null;
  source_type: string;
}

export interface TraceDetail {
  trace_id: string;
  case_id: string;
  user_prompt: string;
  assistant_response: string;
  duration_ms: number | null;
  error: string | null;
  session_id: string | null;
  assessments: AssessmentDetail[];
}

export interface SessionGroup {
  session_id: string;
  eval_type: string;
  traces: TraceDetail[];
  session_assessment: AssessmentDetail | null;
}

export interface TracesResponse {
  traces: TraceDetail[];
  sessions: SessionGroup[];
}

// ---------------------------------------------------------------------------
// Run types
// ---------------------------------------------------------------------------

export interface RunSummary {
  run_id: string;
  timestamp: string;
  params: Record<string, string>;
  metrics: Record<string, number>;
  universal_quality: number | null;
  trace_count: number;
  dataset_id: string | null;
  git_sha: string | null;
}

export interface RunsResponse {
  runs: RunSummary[];
}

// ---------------------------------------------------------------------------
// Experiment types
// ---------------------------------------------------------------------------

export interface ExperimentSummary {
  experiment_id: string;
  name: string;
  eval_type: string;
  run_count: number;
  last_run_timestamp: string | null;
  latest_pass_rate: number | null;
  latest_universal_quality: number | null;
}

export interface ExperimentsResponse {
  experiments: ExperimentSummary[];
}

// ---------------------------------------------------------------------------
// Quality trend types
// ---------------------------------------------------------------------------

export interface QualityTrendPoint {
  eval_type: string;
  timestamp: string;
  universal_quality: number;
  run_id: string;
  git_sha: string;
}

export interface QualityTrendResponse {
  points: QualityTrendPoint[];
}

// ---------------------------------------------------------------------------
// Dataset types
// ---------------------------------------------------------------------------

export interface DatasetCase {
  record_id: string;
  inputs: Record<string, unknown>;
  expectations: Record<string, unknown>;
  extra: Record<string, unknown>;
}

export interface DatasetDetail {
  dataset_id: string;
  name: string;
  dataset_type: string;
  version: string;
  source_file: string;
  case_count: number;
  experiment_ids: string[];
  created_time: string | null;
  cases: DatasetCase[];
}

export interface DatasetsResponse {
  datasets: DatasetDetail[];
}

// ---------------------------------------------------------------------------
// Agent version types
// ---------------------------------------------------------------------------

export interface ExperimentResult {
  experiment_name: string;
  experiment_id: string;
  eval_type: string;
  run_count: number;
  pass_rate: number | null;
  average_quality: number | null;
  latest_run_id: string | null;
}

export interface GuardrailInfo {
  name: string;
  type: string;
}

export interface SpecialistInfo {
  name: string;
  type: string;
  model?: string;
  tools: string[];
  description: string;
}

export interface AgentGraphNode {
  id: string;
  label: string;
  type: string;
  tools?: string[];
}

export interface AgentGraphEdge {
  source: string;
  target: string;
  label: string;
}

export interface AgentGraph {
  nodes: AgentGraphNode[];
  edges: AgentGraphEdge[];
}

export interface AgentConfig {
  model: string;
  name: string;
  framework: string;
  max_tokens: number | null;
  timeout_seconds: number | null;
  system_prompt: string;
  guardrails: GuardrailInfo[];
  specialists: SpecialistInfo[];
  graph: AgentGraph;
}

export interface AgentVersionSummary {
  model_id: string;
  name: string;
  git_branch: string;
  git_commit: string;
  git_commit_short: string;
  creation_timestamp: string;
  aggregate_quality: number | null;
  experiment_count: number;
  total_traces: number;
}

export interface AgentVersionDetail extends AgentVersionSummary {
  git_repo_url: string;
  experiment_results: ExperimentResult[];
  config: AgentConfig;
}

export interface AgentVersionsResponse {
  agents: AgentVersionSummary[];
}

File: src/models/eval_explorer.py
"""Pydantic response models for the eval explorer API."""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Assessment & Trace models
# ---------------------------------------------------------------------------


class AssessmentDetailResponse(BaseModel):
    name: str
    raw_value: Any
    normalized_score: Optional[float] = None
    passed: Optional[bool] = None
    rationale: Optional[str] = None
    source_type: str = ""


class TraceDetailResponse(BaseModel):
    trace_id: str
    case_id: str
    user_prompt: str
    assistant_response: str
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    session_id: Optional[str] = None
    assessments: list[AssessmentDetailResponse]


class SessionGroupResponse(BaseModel):
    session_id: str
    eval_type: str
    traces: list[TraceDetailResponse]
    session_assessment: Optional[AssessmentDetailResponse] = None


class TracesResponse(BaseModel):
    traces: list[TraceDetailResponse]
    sessions: list[SessionGroupResponse]


# ---------------------------------------------------------------------------
# Run models
# ---------------------------------------------------------------------------


class RunSummaryResponse(BaseModel):
    run_id: str
    timestamp: datetime
    params: dict[str, str]
    metrics: dict[str, float]
    universal_quality: Optional[float] = None
    trace_count: int = 0


class RunsResponse(BaseModel):
    runs: list[RunSummaryResponse]


# ---------------------------------------------------------------------------
# Experiment models
# ---------------------------------------------------------------------------


class ExperimentSummaryResponse(BaseModel):
    experiment_id: str
    name: str
    eval_type: str
    run_count: int
    last_run_timestamp: Optional[datetime] = None
    latest_pass_rate: Optional[float] = None
    latest_universal_quality: Optional[float] = None


class ExperimentsResponse(BaseModel):
    experiments: list[ExperimentSummaryResponse]


# ---------------------------------------------------------------------------
# Quality trend models
# ---------------------------------------------------------------------------


class QualityTrendPointResponse(BaseModel):
    eval_type: str
    timestamp: datetime
    universal_quality: float
    run_id: str


class QualityTrendResponse(BaseModel):
    points: list[QualityTrendPointResponse]


# ---------------------------------------------------------------------------
# Dataset models
# ---------------------------------------------------------------------------


class DatasetCaseResponse(BaseModel):
    id: str
    user_prompt: str
    rubric: Optional[str] = None
    tags: list[str] = []
    extra: dict[str, Any] = {}


class DatasetDetailResponse(BaseModel):
    name: str
    file_path: str
    version: str
    description: str
    case_count: int
    cases: list[DatasetCaseResponse] = []


class DatasetsResponse(BaseModel):
    datasets: list[DatasetDetailResponse]
